workshops/views.py

```diff
@@ -3103,10 +3106,131 @@ def duplicates(request):
 
     return render(request, 'workshops/duplicates.html', context)
 
+# ------------------------------------------------------------
+# Views for trainees
+
+def _homework2state(homework):
+    if homework is None:
+        return 'none'
+    elif homework.state == 'n':
+        return 'not-evaluated'
+    elif homework.state == 'p':
+        return 'passed'
+    else:  # failed homework
+        return 'none'
+
+
+@login_required
+def trainee_dashboard(request):
+    swc_form = SendHomeworkForm(submit_name='swc-submit')
+    dc_form = SendHomeworkForm(submit_name='dc-submit')
+
+    # FIXME We hardcode requirement set. What's more, what we do is a bit
+    # different for each Requirement. This will results in problems when we
+    # want to switch to dynamic RequirementSet. Any ideas how to fix that?
```

Comment:
- In the case of Training, Discussion Session, SWC and DC Demo Session, we display whether they passed these requirements (we look for passed, not-discarded TrainingProgress record).

- In the case SWC/DC homework, we look at the state of latest not-discarded TrainingProgress. If it's "failed" or there is no such record, we ask trainee to submit homework. Otherwise, we display "Homework not evaluated yet" or "Homework accepted" depending on the state.

I'm not sure if this answers your question, because I didn't understand what exactly you asked for.